catalogd/cmd/catalogd/main.go

```diff
@@ -74,144 +73,232 @@ const (
+		return err
 	}
 
 	protocol := "http://"
-	if certFile != "" && keyFile != "" {
+	if cfg.certFile != "" && cfg.keyFile != "" {
 		protocol = "https://"
 	}
-	externalAddr = protocol + externalAddr
+	cfg.externalAddr = protocol + cfg.externalAddr
 
-	cfg := ctrl.GetConfigOrDie()
+	k8sCfg := ctrl.GetConfigOrDie()
 
-	cw, err := certwatcher.New(certFile, keyFile)
+	cw, err := certwatcher.New(cfg.certFile, cfg.keyFile)
 	if err != nil {
-		log.Fatalf("Failed to initialize certificate watcher: %v", err)
+		return fmt.Errorf("failed to initialize certificate watcher: %w", err)
 	}
 
 	tlsOpts := func(config *tls.Config) {
 		config.GetCertificate = cw.GetCertificate
-		// Ensure HTTP/2 is disabled by default for webhooks and metrics.
-		// Disabling HTTP/2 mitigates vulnerabilities associated with:
-		// - HTTP/2 Stream Cancellation (GHSA-qppj-fm5r-hxr3)
-		// - HTTP/2 Rapid Reset (GHSA-4374-p667-p6c8)
-		// While CVE fixes exist, they remain insufficient; disabling HTTP/2 helps reduce risks.
-		// For details, see: https://github.com/kubernetes/kubernetes/issues/121197
 		config.NextProtos = []string{"http/1.1"}
 	}
 
-	// Create webhook server and configure TLS
 	webhookServer := crwebhook.NewServer(crwebhook.Options{
-		Port: webhookPort,
-		TLSOpts: []func(*tls.Config){
-			tlsOpts,
-		},
+		Port:    cfg.webhookPort,
+		TLSOpts: []func(*tls.Config){tlsOpts},
 	})
 
-	metricsServerOptions := metricsserver.Options{}
-	if len(certFile) > 0 && len(keyFile) > 0 {
-		setupLog.Info("Starting metrics server with TLS enabled", "addr", metricsAddr, "tls-cert", certFile, "tls-key", keyFile)
+	metricsServerOptions := configureMetricsServer(cfg, tlsOpts)
+
+	cacheOptions := configureCacheOptions(globalPullSecretKey)
+
+	mgr, err := createManager(k8sCfg, cfg, webhookServer, metricsServerOptions, cacheOptions)
+	if err != nil {
+		return err
+	}
+
+	if err := mgr.Add(cw); err != nil {
+		return fmt.Errorf("unable to add certificate watcher: %w", err)
+	}
+
+	if cfg.systemNamespace == "" {
+		cfg.systemNamespace = podNamespace()
+	}
+
+	if err := initializeCacheDirectories(cfg.cacheDir); err != nil {
+		return err
+	}
+
+	unpackCacheBasePath := filepath.Join(cfg.cacheDir, source.UnpackCacheDir)
+	if err := initializeCacheDirectories(cfg.cacheDir); err != nil {
+		return err
+	}
+
+	unpacker := configureUnpacker(cfg.cacheDir, cfg.caCertDir, authFilePath)
+
+	localStorage, err := configureStorage(cfg)
+	if err != nil {
+		return err
+	}
+
+	if err := configureCatalogServer(mgr, cfg, localStorage, cw); err != nil {
+		return err
+	}
+
+	if err := setupControllers(mgr, cfg, unpacker, localStorage, globalPullSecretKey, authFilePath); err != nil {
+		return err
+	}
+
+	if err := setupHealthChecks(mgr); err != nil {
+		return err
+	}
+
+	if err := setupGarbageCollector(mgr, k8sCfg, cfg, unpackCacheBasePath); err != nil {
+		return err
+	}
+
+	if err := setupWebhook(mgr); err != nil {
+		return err
+	}
+
+	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
```

Comment:
Might be nice to call `ctrl.SetupSignalHandler()` from the inlined `Run` command, and then update this `run` function to accept a `context.Context`.